dclock.1
.TH DCLOCK 1 "1 March 1988" "X Version 11"
.SH NAME
dclock - digital clock for X
.SH SYNOPSIS
.B dclock
[-\fItoolkitoption\fP ...] [-option ...]
.SH DESCRIPTION
The
.I dclock 
program displays the time in digital format only.  The time is
updated on a per second basis or on a per minute basis.  This program is
nothing more than a wrapper around the dclock widget not associated with
any particular widget set.
.sp
When the clock is running, the user may change attributes by typing:
.in +2
.ta 2i
.nf
\fBr\fP	Toggles \fBReverse Video\fP.
\fBs\fP	Toggles the \fBseconds\fP display.
\fBb\fP	Toggles the \fBbell\fP attribute.
\fBj\fP	Toggles the \fBjump/scroll\fP attribute.
\fBf\fP	Toggles the \fBfade\fP attribute.
\fBd\fP	Toggles the \fBdate\fP format.
\fBu\fP	Toggles the location of \fBdate\fP (top/bottom).
\fBm\fP	Toggles the \fBmilitary time\fP format.
\fBa\fP	Toggles the \fBalarm clock\fP.
\fBt\fP	Toggles the \fBtails\fP attribute.
\fB:\fP	Toggles the \fBblinking colon\fP.
\fB/\fP	Increases the tilt of the numbers.
\fB\\\fP	Decreases the tilt of the numbers.
\fB+\fP	Increases the thickness of the numbers.
\fB-\fP	Decreases the thickness of the numbers.
\fB>\fP	Increases the spacing between numbers.
\fB<\fP	Decreases the spacing between numbers.
\fBq\fP	quits the program.
.sp
.fi
.in -2
The third mouse button puts the clock in the mode where the alarm clock can
be set.  The alarm must be set in 24-hour format to distringuish between
am and pm time.  Digits can be changed by selecting the digit with the
left or middle mouse button.  The Left button advances the time while the
Middle button moves the time backwards.  Using the left or middle button
over the text at the bottom of the clock toggles the alarm from actually
being set (or, use the 'a' key).
.SH OPTIONS
.I Dclock
accepts all of the standard X Toolkit command line options along with the 
additional options listed below:
.TP 8
.B \-help
This option indicates that a brief summary of the allowed options should be
printed on the standard error.
.TP 8
.B \-bell
.TP 8
.B \-nobell
.br
This option, if enabled, indicates that the bell will beep
once on the half hour and twice on the hour.
.TP 8
.B \-bellFile \fIfilename\fP
This specifies a sound file to use to play the dclock hourly chime.  If
unspecified, or the string "NULL" is used, the system beep is used instead.
The sound file format must be understood by the executable specified by
option "audioPlay".  The "NULL" value is a means to override a system-level
setting to disable external sound playing.
.TP 8
.B \-miltime
.TP 8
.B \-nomiltime
.br
This option, if enabled, causes the clock to display the time in 24 hour
(military) time format.
.TP 8
.B \-utc
.TP 8
.B \-noutc
.br
This option, causes the clock to display the UTC (GMT) time rather than
the default local time.
.TP 8
.B \-tails
.TP 8
.B \-notails
.br
This option places [or removes] "tails" on the digits 6 and 9.
.TP 8
.B \-scroll
.TP 8
.B \-noscroll
.br
When the time changes, the digits scroll from the previous digit to the
next digit.  The -noscroll option turns the behavior off.
.TP 8
.B \-blink
.TP 8
.B \-noblink
.br
The colon separating hours from minutes blinks on and off once per second.
With the -noblink option, it remains on permanently.
.TP 8
.B \-persist
.TP 8
.B \-nopersist
.br
This option determines whether or not the clock display remains in the
reverse video state after the alarm goes off.
.TP 8
.B \-fade
.TP 8
.B \-nofade
.br
This option, when enabled, causes the display to fade from one digit to the
next instead of jumping or scrolling.  If fade is specified, the \fBscroll\fP
attribute will be automatically disabled.
.TP 8
.B \-fadeRate \fIn\fP
The fadeRate attribute takes an integer value which specifies the
number of milliseconds to wait between iterations when fading from
one digit to the next.  A larger value will make the fade slower and
more pronounced.  The default is 50.
.TP 8
.B \-audioPlay \fIfilename\fP
Specifies an executable to use to play audio files for the bell and alarm.
The default is none, which falls back to using the system bell for both the
dclock bell and alarm.
.TP 8
.B \-date \fI"format"\fP
The date is printed under the time in the specified font.  The string
displayed is in the "format" argument.  The format corresponds to the
system call strftime() (see strftime(3)).
.br
.sp
.ti +2
dclock -date "Today is %A"
.sp
Useful examples of strftime() formatting:
.in +2
.nf
%A	Full weekday name
%a	Three-char weekday name (Sun, Mon, Tue...)
%B	Full month name
%b	Three-char month name (Jan, Feb, Mar...)
%d	The date (numerical day number of the month)
%Y	Full year (4 digits)
%y	2-digit year number
.fi
.in -2
.TP 8
.B \-dateup
By default, the date string is at the bottom of the display. This option
starts it at the top, instead.
.TP 8
.B \-alarm
.TP 8
.B \-noalarm
.br
The alarm is turned on or off.  Alarm rings bell and toggles reverse 
video.
.br
.TP 8
.B \-alarmTime \fIHH:MM\fP
If alarm is set, it goes off at time specified in 24-hour format.
.TP 8
.B \-alarmFile \fIfilename\fP
This specifies a sound file to use to play the dclock alarm.  If
unspecified, or the string "NULL" is used, the system beep is used
instead.  The sound file format must be understood by the executable
specified by option "audioPlay".  The "NULL" string is a means to
override a system-level setting to disable external sound playing.
.TP 8
.B \-seconds
This option will update the clock every second and display the time
including the seconds.
.TP 8
.B \-slope \fIvalue\fP
Sets the angle of the digits according to the (floating-point) slope
value \fIvalue\fP.
.TP 8
.B \-thickness \fIwidth_factor\fP
Sets the segment thickness as a (floating-point) ratio to digit width.
.TP 8
.B \-spacing \fIspace_factor\fP
Sets the digit spacing as a (floating-point) ratio to digit width.
.TP 8
.B \-smallsize \fIspace_factor\fP
Sets the seconds digit size as a (floating-point) ratio to the digit size
of the minutes and hours digits.
.TP 8
.B \-second_gap \fIspace_factor\fP
Sets the spacing of the seconds digits from the other digits as a
(floating-point) ratio to digit width.
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.  
The default is ``darkslategray''.
.TP 8
.B \-bd \fIcolor\fP
This option specifies the color to use for the border of the window.
The default is ``black.''
.TP 8
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding the window.
.TP 8
.B \-fg \fIcolor\fP
This option specifies the color to use for displaying text and digit segments.
The default is ``chartreuse''.
.TP 8
.B \-led_off \fIcolor\fP
This option specifies the color to use for displaying digit segments when
turned off.  The default is ``darkgreen''.
.TP 8
.B \-fn \fIfont\fP
This option specifies the font to be used for displaying normal text.  The
default is ``Fixed.''
.TP 8
.B \-geometry \fIgeometry\fP
This option specifies the prefered size and position of the clock window.
.TP 8
.B \-display \fIhost\fP:\fIdisplay\fP
This option specifies the X server to contact.
.TP 8
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.  This is especially
useful for setting resources that do not have separate command line options.
.SH X DEFAULTS
It understands all of the core resource names and
classes as well as:
.PP
.TP 8
.B width (\fPclass\fB Width)
Specifies the width of the clock.
.TP 8
.B height (\fPclass\fB Height)
Specifies the height of the clock.
.TP 8
.B foreground (\fPclass\fB Foreground)
Specifies the color for the digits.  Using the class specifies the
color for all things that normally would appear in the foreground color.
The default is ``chartreuse''.
.TP 8
.B background (\fPclass\fB Background)
Specifies the background color, by default ``DarkSlateGray''.
.TP 8
.B led_off (\fPclass\fB Foreground)
Specifies the color of the digit segments when ``off''.  This can be
set to be the background color, or it can be set to a color just
slightly different from the background, matching the segment colors,
to give just a hint of color.  Default ``DarkGreen'' works well with
the default segment color of ``chartreuse''.
.TP 8
.B bell (\fPclass\fB Boolean)
Specifies whether or not a bell should be rung on the hour and half hour.
.TP 8
.B font (\fPclass\fB Font)
Specifies the font to be used for the date.
.TP 8
.B militaryTime (\fPclass\fB Boolean)
Specifies whether the military (24hr) time format should be used.
.TP 8
.B utc (\fPclass\fB Boolean)
Specifies whether UTC (GMT) time should be used, rather than local time.
.TP 8
.B reverseVideo (\fPclass\fB ReverseVideo)
Specifies that the foreground and background colors should be reversed.
.TP 8
.B scroll (\fPclass\fB Boolean)
Specifies whether the digits should scroll or not.
.TP 8
.B fade (\fPclass\fB Boolean)
Specifies whether the digits should fade or not.
.TP 8
.B fadeRate (\fPclass\fB Time)
The number of milliseconds of delay between iterations of fading.
.TP 8
.B seconds (\fPclass\fB Boolean)
Specifies whether the seconds should be displayed or not.
.TP 8
.B bell (\fPclass\fB Boolean)
Specifies whether the bell should sound on the half hour and on the hour.
.TP 8
.B tails (\fPclass\fB Boolean)
Specifies whether tails should be put on the digits 6 and 9.
.TP 8
.B blink (\fPclass\fB Boolean)
Specifies whether the colon separating hours from minutes should blink on
and off once per second or remain on permanently.
.TP 8
.B alarm (\fPclass\fB Boolean)
Specifies whether the alarm should go off at the specified time (alarmTime).
.TP 8
.B alarmTime (\fPclass\fB String)
Specifies the time alarm goes off if alarm is set.
.TP 8
.B alarmPersist (\fPclass\fB Boolean)
Specifies whether the alarm should remain in reverse video after the alarm
goes off, or return to normal video.
.TP 8
.B angle (\fPclass\fB Float)
Defines the slope of the digits (default 6).
.TP 8
.B smallRatio (\fPclass\fB Float)
Defines the size of the seconds digits compared to the minutes digits
(default 2/3).
.TP 8
.B secondGap (\fPclass\fB Float)
Defines the amount of space between the minutes digits and the seconds digits
(default 0.3).
.TP 8
.B widthFactor (\fPclass\fB Float)
Determines the width of segments as a ratio to digit width (default 0.13).
.TP 8
.B spaceFactor (\fPclass\fB Float)
Determines the spacing between digit as a ratio to digit width (default 0.09).
.TP 8
.B bellFile (\fPclass\fB String)
Specifies an audio file to play for the hourly bell (no default).
.TP 8
.B alarmFile (\fPclass\fB String)
Specifies an audio file to play for the alarm (no default).
.TP 8
.B audioPlay (\fPclass\fB String)
Specifies an executable to use to play the bell and alarm audio files, if
they are specified.  Otherwise, the system bell is used (default is
"/usr/bin/play").
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH "SEE ALSO"
X(1), xrdb(1), time(3C).
.SH BUGS
.I Dclock
believes the system clock.
.PP
Scrolling from 9 to 10 O'Clock seems weird, but chances are you won't
notice it.
.PP
Scrolling can take so long that dclock may lag behind system time by
a fraction of a second and eventually skip a second to catch up.
In modern times, it should be dealing with this by tracking seconds
with a separate thread.
.SH COPYRIGHT
Copyright (c) 1988, Dan Heller.
Revisions copyright (c) 1993 and 1999, Tim Edwards.
.SH AUTHOR
.nf
Original dclock by Dan Heller (1988) <argv@sun.com> or <dheller@cory.berkeley.edu>
.br
Fade & tails by Steve Reinhardt, <steven@fts.webo.dg.com>
.br
Derived geometry by Tim Edwards (1993) <tim@stravinsky.jhuapl.edu>
.br
Antialiasing, color fades, and audio play by Tim Edwards (1999)
<tim@stravinsky.jhuapl.edu>
.fi


debian/changelog
dclock (2.1.2-8) unstable; urgency=low

dclock (2.1.2-8) unstable; urgency=low

  * fix FTBFS (X11's imake build procedure is gone).  Thanks to Harri
    Dunkel for the patch (closes: #363055)

 -- Adrian Bridgett <bridgett@debian.org>  Sun,  9 Jul 2006 09:50:08 +0100

dclock (2.1.2-7) unstable; urgency=low

  * allow overriding of bellFile sound and alarmFile sounds. 
    Based on a patch by Nadim Shaikli (closes: #289780)
  * upgrade DH_COMPAT to v4

 -- Adrian Bridgett <bridgett@debian.org>  Mon, 29 Aug 2005 14:06:08 +0100

dclock (2.1.2-6) unstable; urgency=low

  * ACK of NMU
  * disable external sounds if we have problems using them (closes: #266359)
  * only ring the alarm bell once, not as many times as we can in 5seconds

 -- Adrian Bridgett <bridgett@debian.org>  Tue,  7 Sep 2004 10:46:11 +0100

dclock (2.1.2-5.1) unstable; urgency=low

  * Non-maintainer upload
  * Dclock.c, playbell(): print error and exit if audio player can not
    be execv'ed  (closes: #253704)
  * recommend package 'sox' to enable sound and explain it in the package
    description (closes: #234258)
  * build-depend on debhelper >= 2 (closes: #259729)

 -- Martin Pitt <mpitt@debian.org>  Fri, 16 Jul 2004 14:26:56 +0200

dclock (2.1.2-5) unstable; urgency=low

  * manpage didn't mention -noblink option (closes: #227068)
    thanks to Sven Rudolph for spotting this and providing a patch.

 -- Adrian Bridgett <bridgett@debian.org>  Sun, 18 Jan 2004 13:44:46 +0000

dclock (2.1.2-4) unstable; urgency=low

  * s/miltime/militaryTime/ for X resource in manpage (closes: #198669)

 -- Adrian Bridgett <bridgett@debian.org>  Tue, 24 Jun 2003 21:37:37 +0100

dclock (2.1.2-3) unstable; urgency=low

  * lintian fixes
  * darn typo in last changelog - let's close it here (closes: #185864)

 -- Adrian Bridgett <bridgett@debian.org>  Sat, 31 May 2003 22:01:50 +0100

dclock (2.1.2-2) unstable; urgency=low

  * patch from Robert Macomber to stop dclock missing some updates.
    I've actually only added 1 onto next_intr, as I can explain that, but
    not 100 (close: #185864)

 -- Adrian Bridgett <bridgett@debian.org>  Mon, 21 Apr 2003 19:02:40 +0100

dclock (2.1.2-1) unstable; urgency=low

  * slightly newer code at upstream location (I've named it 2.1.2)
  * build-depends were right, mark bug as fixed (closes: #122166)
  * app-defaults file was wrongly named. Set dclock.input:true to fix
    a WM interaction(closes: #98840)

 -- Adrian Bridgett <bridgett@debian.org>  Sat,  2 Feb 2002 21:26:38 +0000

dclock (2.1.1-6) unstable; urgency=low

  * add some build-depends for #122166

 -- Adrian Bridgett <bridgett@debian.org>  Mon,  3 Dec 2001 18:53:36 +0000

dclock (2.1.1-5) unstable; urgency=low

  * refer to the X Window System properly (close: #119841)

 -- Adrian Bridgett <bridgett@debian.org>  Sun,  2 Dec 2001 14:32:00 +0000

dclock (2.1.1-4) unstable; urgency=low

  * app-defaults files must be marked conffiles

 -- Adrian Bridgett <bridgett@debian.org>  Sun,  4 Mar 2001 18:55:21 +0000

dclock (2.1.1-3) unstable; urgency=low

  * use time.h not sys/time.h (closes: #86957)

 -- Adrian Bridgett <bridgett@debian.org>  Sat,  3 Mar 2001 15:33:23 +0000

dclock (2.1.1-2) unstable; urgency=low

  * use new app-defaults location (closes: #86265)

 -- Adrian Bridgett <bridgett@debian.org>  Sun, 18 Feb 2001 20:31:10 +0000

dclock (2.1.1-1) unstable; urgency=low

  * new upstream (thanks to Harald Dunkel)
    NB: The version number _hadn't_ changed upstream
  * use /usr/share/sounds like other programs (closes: #79159)

 -- Adrian Bridgett <bridgett@debian.org>  Tue, 26 Dec 2000 18:50:50 +0000

dclock (2.1-1) unstable; urgency=low

  * initial release.

 -- Adrian Bridgett <bridgett@debian.org>  Mon,  3 Jan 2000 14:47:11 +0000


dclock.c
/*
 * Copyright (c) 1988 Dan Heller <argv@sun.com>
 * dclock -- program to demonstrate how to use the digital-clock widget.
 * To specify a date, the date format is a string corresponding to the
 * syntax of strftime(3).
 * To specify seconds to be displayed, use "-seconds" or use the resource
 * manager: *Dclock.seconds: on
 */
#include <stdio.h>
#include <locale.h>
#include <X11/Intrinsic.h>
#include "Dclock.h"

static XrmOptionDescRec options[] = {
    {"-date",	   "*Dclock.date",	  XrmoptionSepArg, NULL    },
    {"-dateup",    "*Dclock.dateUp",	  XrmoptionNoArg,  "TRUE"  },
    {"-nodateup",  "*Dclock.dateUp",	  XrmoptionNoArg,  "FALSE" },
    {"-seconds",   "*Dclock.seconds",	  XrmoptionNoArg,  "TRUE"  },
    {"-miltime",   "*Dclock.militaryTime",XrmoptionNoArg,  "TRUE"  },
    {"-nomiltime", "*Dclock.militaryTime",XrmoptionNoArg,  "FALSE" },
    {"-stdtime",   "*Dclock.militaryTime",XrmoptionNoArg,  "FALSE" },
    {"-utc",       "*Dclock.utcTime",     XrmoptionNoArg,  "TRUE"  },
    {"-noutc",     "*Dclock.utcTime",     XrmoptionNoArg,  "FALSE" },
    {"-disptime",  "*Dclock.displayTime", XrmoptionNoArg,  "TRUE"  },
    {"-audioPlay", "*Dclock.audioPlay",	  XrmoptionSepArg, NULL    },
    {"-bell",	   "*Dclock.bell",	  XrmoptionNoArg,  "TRUE"  },
    {"-nobell",	   "*Dclock.bell",	  XrmoptionNoArg,  "FALSE" },
    {"-bellFile",  "*Dclock.bellFile",	  XrmoptionSepArg, NULL    },
    {"-alarm",     "*Dclock.alarm",	  XrmoptionNoArg,  "TRUE"  },
    {"-noalarm",   "*Dclock.alarm",	  XrmoptionNoArg,  "FALSE" },
    {"-persist",   "*Dclock.alarmPersist",XrmoptionNoArg,  "TRUE"  },
    {"-nopersist", "*Dclock.alarmPersist",XrmoptionNoArg,  "FALSE" },
    {"-alarmTime", "*Dclock.alarmTime",	  XrmoptionSepArg, NULL    },
    {"-alarmFile", "*Dclock.alarmFile",	  XrmoptionSepArg, NULL    },
    {"-blink",     "*Dclock.blink",	  XrmoptionNoArg,  "TRUE"  },
    {"-noblink",   "*Dclock.blink",	  XrmoptionNoArg,  "FALSE" },
    {"-tails",     "*Dclock.tails",	  XrmoptionNoArg,  "TRUE"  },
    {"-notails",   "*Dclock.tails",	  XrmoptionNoArg,  "FALSE" },
    {"-scroll",    "*Dclock.scroll",	  XrmoptionNoArg,  "TRUE"  },
    {"-noscroll",  "*Dclock.scroll",	  XrmoptionNoArg,  "FALSE" },
    {"-fade",	   "*Dclock.fade",	  XrmoptionNoArg,  "TRUE"  },
    {"-nofade",	   "*Dclock.fade",	  XrmoptionNoArg,  "FALSE" },
    {"-fadeRate",  "*Dclock.fadeRate",	  XrmoptionSepArg, NULL    },
    {"-slope",	   "*Dclock.angle",	  XrmoptionSepArg, NULL    },
    {"-thickness", "*Dclock.widthFactor", XrmoptionSepArg, NULL    },
    {"-smallsize", "*Dclock.smallRatio",  XrmoptionSepArg, NULL    },
    {"-spacing",   "*Dclock.spaceFactor", XrmoptionSepArg, NULL    },
    {"-second_gap","*Dclock.secondGap",	  XrmoptionSepArg, NULL    },
    {"-led_off",   "*Dclock.led_off",	  XrmoptionSepArg, NULL    },
#ifdef XFT_SUPPORT
    {"-fontName",  "*Dclock.fontName",	  XrmoptionSepArg, NULL    },
#endif
};

static void
Usage(name, args)
String name, *args;
{
    static char *help_message[] = {
	"where options include:",
	"    -help			print this help text",
	"    -bg color			field background color",
	"    -fg color			segment foreground color",
	"    -fn font			font name",
	"    -geometry geom		size of mailbox",
	"    -display host:dpy		X server to contact",
	"    -led_off color		segment background color",
	"    -date \"date format\"	show the date in strftime(3) format",
	"    -[no]dateup 		[don't]	put the date up at the top",
	"    -[no]seconds 		[don't]	display seconds",
	"    -[no]miltime 		[don't]	display time in 24 hour format",
	"    -[no]utc			[don't] display the UTC time instead of local",
	"    -[no]blink			[don't] blink the colon",
	"    -[no]scroll 		turn on [off] scrolling of numbers",
	"    -[no]tails			draw [remove] tails on digits 6 and 9",
	"    -[no]fade			[don't] fade numbers",
	"    -[no]bell 			[don't] ring bell each half hour",
	"    -[no]persist		[don't] leave in reverse video after alarm",
	"    -bellFile filename		sound file for bell sound",
	"    -[no]alarm 		turn on/off alarm",
	"    -alarmTime hh:mm		time alarm goes off",
	"    -alarmFile filename	sound file for alarm sound",
        "    -audioPlay	filename	executable to use to play bell and alarm",
	"    -fadeRate int_val		wait between fade steps (in msec)",
	"    -slope float_val		set angle of the digits",
	"    -smallsize float_val	set size ratio of the small to large digits",
	"    -second_gap float_val	set spacing between minutes and seconds digits",
	"    -thickness	float_val	set segment thickness as ratio to digit width",
	"    -spacing float_val		set digit spacing as ratio to digit width",
#ifdef XFT_SUPPORT
	"    -fontName name		name of freefont font to use for date string",
#endif
	NULL
    };
    char **cpp;

    fprintf(stderr, "Invalid Args:");
    while (*args)
	fprintf(stderr, " \"%s\"", *args++);
    fputc('\n', stderr);
    fprintf(stderr, "usage: %s [-options ...]\n", name);
    for (cpp = help_message; *cpp; cpp++)
	fprintf(stderr, "%s\n", *cpp);
    exit(1);
}

static void
quit()
{
    exit(0);
}

static XtActionsRec actionsList[] = {
    { "quit",	quit },
};

main(argc, argv)
char *argv[];
{
    XtAppContext app;
    Widget toplevel, clock_w;
    char *name, *rindex();
    XWMHints     *wmhints;       /* for proper input focus */

    if (name = rindex(argv[0], '/'))
	name++;
    else
	name = argv[0];

    setlocale(LC_TIME, "");

    toplevel = XtAppInitialize(&app, "Dclock", options, XtNumber(options),
			&argc, argv, (String *)NULL, (ArgList)NULL, 0);
    XtAppAddActions(app, actionsList, 1);

    if (argc != 1)
	Usage(name, argv+1);

    clock_w = XtCreateManagedWidget(name, dclockWidgetClass, toplevel, NULL, 0);
    XtOverrideTranslations(clock_w, XtParseTranslationTable("<Key>q: quit()"));

    XtRealizeWidget(toplevel);

    wmhints = XGetWMHints(XtDisplay(toplevel), XtWindow(toplevel));
    wmhints->input = True;
    wmhints->flags |= InputHint;
    XSetWMHints(XtDisplay(toplevel), XtWindow(toplevel), wmhints);
    XFree(wmhints);
    XtAppMainLoop(app);
    return 0;
}
